heft-plugins/heft-webpack4-plugin/src/Webpack4Plugin.ts

```diff
@@ -0,0 +1,292 @@
+  IWebpackConfiguration,
+  IWebpackConfigurationWithDevServer,
+  IWebpack4PluginAccessor
+} from './shared';
+import { WebpackConfigurationLoader } from './WebpackConfigurationLoader';
+
+export interface IWebpack4PluginOptions {
+  devConfigurationPath: string | undefined;
+  configurationPath: string | undefined;
+}
+
+/**
+ * @public
+ */
+export const PLUGIN_NAME: 'Webpack4Plugin' = 'Webpack4Plugin';
+const WEBPACK_DEV_SERVER_PACKAGE_NAME: 'webpack-dev-server' = 'webpack-dev-server';
+const WEBPACK_DEV_SERVER_ENV_VAR_NAME: 'WEBPACK_DEV_SERVER' = 'WEBPACK_DEV_SERVER';
+const UNINITIALIZED: 'UNINITIALIZED' = 'UNINITIALIZED';
+
+/**
+ * @internal
+ */
+export default class Webpack4Plugin implements IHeftTaskPlugin<IWebpack4PluginOptions> {
+  private _webpack: typeof TWebpack | undefined;
+  private _webpackConfiguration: IWebpackConfiguration | undefined | typeof UNINITIALIZED = UNINITIALIZED;
+
+  public readonly accessor: IWebpack4PluginAccessor = {
+    hooks: {
+      onLoadConfiguration: new AsyncSeriesBailHook(),
+      onConfigure: new AsyncSeriesHook(['webpackConfiguration']),
+      onAfterConfigure: new AsyncParallelHook(['webpackConfiguration']),
+      onEmitStats: new AsyncParallelHook(['webpackStats'])
+    }
+  };
+
+  public apply(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    options: IWebpack4PluginOptions
+  ): void {
+    // These get set in the run hook and used in the onConfigureWebpackHook
+    let production: boolean;
+    let serveMode: boolean;
+    let watchMode: boolean;
+
+    // Provide the initial webpack configuration by tapping the hook. Tap with stage as
+    // MAX_SAFE_INTEGER to ensure that we run last.
+    this.accessor.hooks.onLoadConfiguration.tapPromise(
+      { name: PLUGIN_NAME, stage: Number.MAX_SAFE_INTEGER },
+      async () => {
+        taskSession.logger.terminal.writeVerboseLine('Loading default Webpack configuration');
+        const configurationLoader: WebpackConfigurationLoader = new WebpackConfigurationLoader(
+          taskSession.logger,
+          production,
+          serveMode
+        );
+        const webpack: typeof TWebpack = await this._loadWebpackAsync();
+        return await configurationLoader.tryLoadWebpackConfigurationAsync({
+          ...options,
+          taskSession,
+          heftConfiguration,
+          webpack
+        });
+      }
+    );
+
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Obtain the finalized webpack configuration
+      const webpackConfiguration: IWebpackConfiguration | undefined =
+        await this._getWebpackConfigurationAsync();
+      if (webpackConfiguration) {
+        const webpackConfigurationArray: IWebpackConfigurationWithDevServer[] = Array.isArray(
+          webpackConfiguration
+        )
+          ? webpackConfiguration
+          : [webpackConfiguration];
+
+        // Add each output path to the clean list
```

Comment:
Add an express comment that if webpack plugins write assets to paths that start with `../` (which you can, in fact, do, and some of our plugins use) they will need to be manually added to the cleanup list.